# Sample Size Estimation: Two-Sample Mean Comparison

In [2]:
import numpy as np
import pandas as pd
import statsmodels.formula.api as smf
from scipy.stats import norm
import matplotlib.pyplot as plt


## Background

The school wants to effectively improve student academic achievement. They plan to launch a new initiative program requiring or encouraging students to actively participate in certain extracurricular activities (Initiative B), with the hope that extracurricular activities will foster discipline and teamwork, which will positively impact test scores.

Before implementing this policy across the school, teachers conducted a controlled experiment to compare two groups:

- Group A (Control): Students who focused solely on classwork and did not participate in extracurricular activities.
- Group B (Treatment): Students who actively participated in extracurricular activities.

Assumptions & Known Parameters:
- Based on historical exam data from one month ago, the standard deviation (σ) of test scores is 15 points.
- The desired difference in mean scores (effect size δ) is 5 points.
- The available dataset includes 4,234 students (2,287 who did not participate in extracurricular activities, 1,947 who participated).

The main question is: **what is the minimum sample size needed** for the results of this experiment to be statistically reliable?

---
***Helper Functions***

There are two approaches to estimating sample size:

### 1. Based on Standard Error (SE)

Used when we want to limit the **precision of the estimate** — how close the estimated interval between the two means is.

$$
n=\left (\frac{2\sigma}{\text{SE}(\bar{y}_1-\bar{y}_2)}  \right )^2
$$

  - $n$ = total sample size required
  - $\sigma$ = standard deviation of population
  - $\text{SE}(\bar{y}_1-\bar{y}_2)$ = desired standard error of the difference between two means

In [3]:
def sample_size_diff_mean_se(sigma, standard_error):
    
    # Calculate sample size
    n = ( (2 * sigma) / standard_error )**2
    
    return n

### 2. Based on Desired Power dan Confidence Level

Used when we want to guarantee the **ability to detect real effects** with a certain level of confidence.

$$
n = \frac{(2\sigma(z_{1-{\frac{\alpha}{2}}}+z_{1-\beta}))^2}{\delta^2}
$$

  - $n$ = total sample size required (two combined groups)
  - $\sigma$ = standard deviation of population
  - $\delta$ = desired difference of two mean to be detected
  - $z_{1-\frac{\alpha}{2}}$ = z-value based on significance level ($\alpha$), or 1 - confidence level
  - $z_{1-\beta}$ = z-value based on desired power ($1-\beta$)

In [4]:
def sample_size_diff_mean(alpha, beta, effect_size, sigma):
    
    # Calculate z-value
    z_alpha = norm.ppf(1 - alpha/2)
    z_beta = norm.ppf(1 - beta)
    
    # Calculate sample size
    n =  (2 * sigma * (z_alpha+z_beta))**2 / (effect_size**2)
    
    return n

---
### Estimate 1: Based on Standard Error

**Question:** What is the estimated sample size needed to infer the difference in mean test scores between two groups with a **standard error = 2 points**?

This approach is suitable when we want to ensure that the estimate of the mean difference is sufficiently precise — in this case, the uncertainty (SE) is no more than 2 points.

**Parameters:**
- σ = 13 points (estimated from pilot or historical data)
- Desired SE = 2 points

In [5]:
# Given
standard_error = 2
sigma = 13

In [6]:
# Calculate the number of sample size
n = sample_size_diff_mean_se(sigma, standard_error)

print(f"Number of total sample size needed            : {n:.2f}")
print(f"Number of sample size of each group needed    : {n/2:.2f}")

Number of total sample size needed            : 169.00
Number of sample size of each group needed    : 84.50


**Estimation Conclusion 1:**

To obtain an estimate of the difference in mean test scores with an SE of ≤ 2 points**, the following are required:
- **Total: 169 samples** (rounded up)
- **Per group: 85 samples** (assuming two balanced groups)

> ✅ **Important note:** The available dataset has 4,234 students — far exceeding this minimum requirement. This means that our analysis will produce a **very precise** estimate, with an actual SE well below 2 points.

---
## Estimation 2: Based on Power and Confidence Level

Question: What is the estimated sample size if we want the experiment to have 80% power to detect a 5-point difference between groups, with a 95% confidence level?

This approach is more commonly used in experimental design. We define:
- 95% confidence level → we are willing to accept a 5% risk of incorrectly rejecting a true H₀ (Type I error/false positive)
- 80% power → we want an 80% probability of successfully detecting a true effect of δ = 5 points (avoiding a Type II error/false negative)

Parameters:
- Effect size (δ) = 5 points
- Confidence level = 95%
- Power = 80%
- σ = 15 points (from historical data)

In [7]:
# Given
effect_size = 5    # absolute desired difference of two groups means
conf = 0.95        # confidence level
power = 0.8        # statistical power
sigma = 15         # standard deviation of population (from historical data)

In [8]:
# Calculate alpha (significance level) and beta (type II error rate)
beta = 1 - power
alpha = 1 - conf

In [9]:
# Calculate the number of total sample size
n = sample_size_diff_mean(alpha, beta, effect_size, sigma)

print(f"Number of total sample size needed            : {n:.2f}")
print(f"Number of sample size of each group needed    : {n/2:.2f}")

Number of total sample size needed            : 282.56
Number of sample size of each group needed    : 141.28


**Estimation Conclusion 2:**

To detect a 5-point difference between groups with **95% confidence and 80% power**, the following are required:
- **Total: 283 samples** (rounded up, or ~300 as a safe number)
- **Per group: 142 samples** (or ~150 as a safe number)

> ✅ **Important note:** The available dataset contains 1,947 extracurricular students and 2,287 non-excursion students, **far exceeding the minimum requirement** of 142 per group. This means our experiment has very high power (approaching 100%), and even small differences will be statistically detectable.

---
## Sensitivity Analysis: The Effect of Changes in Alpha & Beta

The following is a sensitivity analysis to see how changes in the significance level (α) and power (β) affect sample size requirements. This is useful for understanding trade-offs in experimental design:

- **Scenario A:** α = 0.10 (90% confidence), power = 80% → more lenient, requiring a smaller sample size
- **Scenario B:** α = 0.50 (50% confidence), power = 95% → very high power, dominant power effect

In [10]:
# Scenario A: Confidence level is lowered to 90% (alpha=0.10), power remains at 80%
print("=== Skenario A: alpha=0.10, power=80% ===")
n = sample_size_diff_mean(alpha=0.1, beta=0.2, effect_size=5, sigma=15)
print(f"Number of total sample size needed            : {n:.2f}")
print(f"Number of sample size of each group needed    : {n/2:.2f}")

=== Skenario A: alpha=0.10, power=80% ===
Number of total sample size needed            : 222.57
Number of sample size of each group needed    : 111.29


In [11]:
# Scenario B: Confidence is lowered to 50% (alpha=0.5), power is increased to 95%
print("=== Skenario B: alpha=0.50, power=95% ===")
n = sample_size_diff_mean(alpha=0.5, beta=0.05, effect_size=5, sigma=15)
print(f"Number of total sample size needed            : {n:.2f}")
print(f"Number of sample size of each group needed    : {n/2:.2f}")

=== Skenario B: alpha=0.50, power=95% ===
Number of total sample size needed            : 193.66
Number of sample size of each group needed    : 96.83


---
## Final Summary

| Method | Parameters | Total Sample | Per Group |
|--------|-----------|:------------:|:------------:|
| SE-based | σ=13, SE=2 | ~170 | ~85 |
| Power-based (main) | σ=15, δ=5, conf=95%, power=80% | ~283 (~300) | ~142 (~150) |
| Sensitivity A | conf=90%, power=80% | ~223 | ~112 |
| Sensitivity B | conf=50%, power=95% | ~194 | ~97 |

**Interpretation:**

The estimated sample size (~142–300 samples) is the **minimum requirement** for valid statistical inference. This does not mean that this is the only size that can be used** — quite the opposite.

The available dataset of 4,234 students** (well above the minimum) offers the following advantages:
1. **Statistical power approaching 100%** — even the smallest significant differences are detected.
2. **Very small standard error** — estimates of mean differences are very precise.
3. **Better representation** — subgroup analyses (by gender, region, etc.) remain valid.

> 💡 **Conclusion:** With 4,234 available data sets, this comparative analysis of test scores between extracurricular and non-excursion groups has a **very strong** statistical foundation to support school policy decisions.